chromium_src/chrome/browser/notifications/notification_display_service_impl.cc

```diff
@@ -0,0 +1,130 @@
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "build/build_config.h"
+
+#if defined(OS_ANDROID)
+#include "chrome/browser/notifications/notification_display_service_impl.h"
+
+#define ProcessNotificationOperation ProcessNotificationOperation_ChromiumImpl
+#include "../../../../../../chrome/browser/notifications/notification_display_service_impl.cc"
+#undef ProcessNotificationOperation
+
+#include "base/android/application_status_listener.h"
+#include "chrome/browser/lifetime/application_lifetime_android.h"
+
+namespace {
+// Wait before shutdown if running headless (seconds)
+const int64_t kWaitBeforeShutdownWhenRunHeadless = 60;
+}  // namespace
+
+void NotificationDisplayServiceImpl::BraveOperationCompleted(
+    const std::string& notification_id) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  auto state = base::android::ApplicationStatusListener::GetState();
+  bool no_ui =
+      (state == base::android::ApplicationState::APPLICATION_STATE_UNKNOWN ||
+       state == base::android::ApplicationState::
+                    APPLICATION_STATE_HAS_DESTROYED_ACTIVITIES);
+
+  if (no_ui && last_dismissed_notification_id_ == notification_id) {
+    last_dismissed_notification_id_ = "";
+    // Wait for 60 seconds and close the browser
+    no_ui_shutdown_timer_.Start(
+        FROM_HERE,
+        base::TimeDelta::FromSeconds(kWaitBeforeShutdownWhenRunHeadless), this,
+        &NotificationDisplayServiceImpl::ShutdownTimerCallback);
+  }
+}
+
+void NotificationDisplayServiceImpl::ShutdownTimerCallback() {
+  // Check if browser runs without UI and close it if so. If during
+  // shutdown a new notification event arrives, timer is stopped in
+  // ProcessNotificationOperation
+  auto state = base::android::ApplicationStatusListener::GetState();
+  bool no_ui =
+      (state == base::android::ApplicationState::APPLICATION_STATE_UNKNOWN ||
+       state == base::android::ApplicationState::
+                    APPLICATION_STATE_HAS_DESTROYED_ACTIVITIES);
+  if (no_ui) {
+    chrome::TerminateAndroid();
+  }
+}
+
+void NotificationDisplayServiceImpl::ProcessNotificationOperation(
+    NotificationCommon::Operation operation,
+    NotificationHandler::Type notification_type,
+    const GURL& origin,
+    const std::string& notification_id,
+    const base::Optional<int>& action_index,
+    const base::Optional<base::string16>& reply,
+    const base::Optional<bool>& by_user) {
+  NotificationHandler* handler = GetNotificationHandler(notification_type);
+  DCHECK(handler);
+  if (!handler) {
+    LOG(ERROR) << "Unable to find a handler for "
+               << static_cast<int>(notification_type);
+    return;
+  }
+
+  no_ui_shutdown_timer_.Stop();
+
+  base::OnceClosure completed_closure =
+      base::BindOnce(&NotificationDisplayServiceImpl::BraveOperationCompleted,
+                     base::Unretained(this), notification_id);
+  base::OnceClosure notification_handler;
+
+  switch (operation) {
```

Comment:
we definitely do not want to duplicate this code and this seems like it would be applied to all notifications, not just brave ads